### PR TITLE
test: simulate for worker node heartbeat timeout

### LINKED ISSUE #7626: simulation test: support simulate worker node heartbeat timeout
### Is your feature request related to a problem? Please describe.

The simulator already supports [kill_node](https://github.com/risingwavelabs/risingwave/blob/db4f27e56a09d234de074a3acbb43bfea70825af/src/tests/simulation/src/cluster.rs#L367), but it always restarts the worker immediately before worker heartbeat timeout, thus the worker rejoins [with the original identify](https://github.com/risingwavelabs/risingwave/blob/db4f27e56a09d234de074a3acbb43bfea70825af/src/meta/src/manager/cluster.rs#L136)(not expired). That is, we haven't covered cases like this: 
1. The worker crashes.
2. Then the worker is removed as expired from cluster due to heartbeat timeout.
3. Then the worker rejoins the cluster and is recognized [with a new identify](https://github.com/risingwavelabs/risingwave/blob/db4f27e56a09d234de074a3acbb43bfea70825af/src/meta/src/manager/cluster.rs#L137).

This kind of cases covers [considerable code logic](https://github.com/risingwavelabs/risingwave/blob/db4f27e56a09d234de074a3acbb43bfea70825af/src/meta/src/barrier/recovery.rs#L129).

### Describe the solution you'd like

_No response_

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by wangrunji0408**
Signed-off-by: Runji Wang <wangrunji0408@163.com>

I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

This PR resolves #7626. It introduces a long delay with a small chance after a node is killed. This will cover the case that the node is expired and removed from the cluster in recovery.

The test failed after this change and I am looking for the reason.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
#7626

### CODE REVIEW

File: src/tests/simulation/src/cluster.rs [hunk 1]
```diff
@@ -421,7 +421,14 @@ impl Cluster {
...
             madsim::runtime::Handle::current().kill(name);
 
-            let t = rand::thread_rng().gen_range(Duration::from_secs(0)..Duration::from_secs(1));
+            let mut t =
+                rand::thread_rng().gen_range(Duration::from_secs(0)..Duration::from_secs(1));
+            // has a small chance to restart after a long time
+            // so that the node is expired and removed from the cluster
+            if rand::thread_rng().gen_bool(0.1) {
+                // max_heartbeat_interval_secs = 60
+                t += Duration::from_secs(120);
```
Comment (jon-chuang) [inline src/tests/simulation/src/cluster.rs:430]: This will affect only time, right?

Comment (wangrunji0408) [reply src/tests/simulation/src/cluster.rs:430]: Sure. It goes fast in simulation.

**Comment by wangrunji0408** [comment]
Found a bug in madsim that it panics when a lease is not found. It should return an error instead. Fixed in https://github.com/madsim-rs/madsim/pull/112.

**Comment by wangrunji0408** [comment]
Found a panic in the case:
1. kill meta-{1,2,3}, compute-1
2. restart meta-{2,3} soon, but meta-1 is keeping offline
3. compute-1 panicked after ~20s:
```
thread ‘<unnamed>’ panicked at ‘called `Result::unwrap()` on an `Err` value: 
GrpcStatus(Status { code: Unavailable, message: “connection refused”, source: None })’, 
src/compute/src/server.rs:90:6
```

cc @shanicky

File: src/tests/simulation/src/cluster.rs [hunk 2]
```diff
@@ -438,7 +442,14 @@ impl Cluster {
             tracing::info!("kill {name}");
             madsim::runtime::Handle::current().kill(name);
 
-            let t = rand::thread_rng().gen_range(Duration::from_secs(0)..Duration::from_secs(1));
+            let mut t =
+                rand::thread_rng().gen_range(Duration::from_secs(0)..Duration::from_secs(1));
+            // has a small chance to restart after a long time
+            // so that the node is expired and removed from the cluster
```
Comment (jon-chuang) [inline src/tests/simulation/src/cluster.rs:448]: I guess the node will be assigned a new worker id?

Comment (wangrunji0408) [reply src/tests/simulation/src/cluster.rs:448]: Yes

**Comment by yezizp2012** [comment]
The previous failure is a little bit confusing 🥵 , may related to some retry interval settings both in kernel and testing side. I will test it fully in local and merge it if there are no problems.

File: src/tests/simulation/src/risingwave.toml [hunk 3]
```diff
@@ -2,6 +2,10 @@
 #
 # Note: this file is embedded in the binary and cannot be changed without recompiling.
 
+[meta]
+# a relatively small number to make it easier to timeout
+max_heartbeat_interval_secs = 10
```
Comment (CAJan93) [inline src/tests/simulation/src/risingwave.toml:7]: Should we maybe go smaller? We use have a timeout of 10s for a meta leader lease in the [example.toml](https://github.com/risingwavelabs/risingwave/blob/43c81b9275d70167ccc16055a09fdece5f6e3d1b/src/config/example.toml#L17). If we use that here we may not get a timeout.

**Review by yezizp2012 — APPROVED** [review]

**Review by shanicky — APPROVED** [review]
After implementing the fixes in this PR, we successfully ran 1500 recovery tests without any failures. Therefore, I suggest that we merge this PR first and then make further modifications to apply `max_heartbeat_interval_secs` from meta to recovery tests for more robust testing.

**Pull request merged**